File: src/main/java/org/jetbrains/kotlinx/lincheck/CTestConfiguration.java
package org.jetbrains.kotlinx.lincheck;

/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import kotlin.jvm.functions.Function0;
import org.jetbrains.kotlinx.lincheck.execution.*;
import org.jetbrains.kotlinx.lincheck.strategy.*;
import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingCTest;
import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingCTestConfiguration;
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressCTest;
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressCTestConfiguration;
import org.jetbrains.kotlinx.lincheck.verifier.Verifier;
import org.jetbrains.kotlinx.lincheck.verifier.linearizability.LinearizabilityVerifier;

import java.lang.reflect.*;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.function.UnaryOperator.identity;
import static org.jetbrains.kotlinx.lincheck.UtilsKt.chooseSequentialSpecification;
import static org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingCTestConfiguration.DEFAULT_GUARANTEES;

/**
 * Configuration of an abstract concurrent test.
 * Should be overridden for every strategy.
 */
public abstract class CTestConfiguration {
    public static final int DEFAULT_ITERATIONS = 200;
    public static final int DEFAULT_THREADS = 2;
    public static final int DEFAULT_ACTORS_PER_THREAD = 5;
    public static final int DEFAULT_ACTORS_BEFORE = 5;
    public static final int DEFAULT_ACTORS_AFTER = 5;
    public static final Class<? extends ExecutionGenerator> DEFAULT_EXECUTION_GENERATOR = RandomExecutionGenerator.class;
    public static final Class<? extends Verifier> DEFAULT_VERIFIER = LinearizabilityVerifier.class;
    public static final boolean DEFAULT_MINIMIZE_ERROR = true;
    public static final long DEFAULT_TIMEOUT_MS = 10_000;

    public final Class<?> testClass;
    public final int iterations;
    public final int threads;
    public final int actorsPerThread;
    public final int actorsBefore;
    public final int actorsAfter;
    public final Class<? extends ExecutionGenerator> generatorClass;
    public final Class<? extends Verifier> verifierClass;
    public final boolean requireStateEquivalenceImplCheck;
    public final Boolean minimizeFailedScenario;
    public final Class<?> sequentialSpecification;
    public final long timeoutMs;

    protected CTestConfiguration(Class<?> testClass, int iterations, int threads, int actorsPerThread, int actorsBefore, int actorsAfter,
        Class<? extends ExecutionGenerator> generatorClass, Class<? extends Verifier> verifierClass,
        boolean requireStateEquivalenceImplCheck, boolean minimizeFailedScenario, Class<?> sequentialSpecification, long timeoutMs)
    {
        this.testClass = testClass;
        this.iterations = iterations;
        this.threads = threads;
        this.actorsPerThread = actorsPerThread;
        this.actorsBefore = actorsBefore;
        this.actorsAfter = actorsAfter;
        this.generatorClass = generatorClass;
        this.verifierClass = verifierClass;
        this.requireStateEquivalenceImplCheck = requireStateEquivalenceImplCheck;
        this.minimizeFailedScenario = minimizeFailedScenario;
        this.sequentialSpecification = sequentialSpecification;
        this.timeoutMs = timeoutMs;
    }

    protected abstract Strategy createStrategy(Class<?> testClass, ExecutionScenario scenario,
                                               List<Method> validationFunctions, Method stateRepresentation, Verifier verifier);

    static List<CTestConfiguration> createFromTestClassAnnotations(Class<?> testClass) {
        Stream<CTestConfiguration> stressConfigurations = Arrays.stream(testClass.getAnnotationsByType(StressCTest.class))
            .map(ann -> new StressCTestConfiguration(testClass, ann.iterations(),
                    ann.threads(), ann.actorsPerThread(), ann.actorsBefore(), ann.actorsAfter(),
                    ann.generator(), ann.verifier(), ann.invocationsPerIteration(), true,
                    ann.requireStateEquivalenceImplCheck(), ann.minimizeFailedScenario(),
                    chooseSequentialSpecification(ann.sequentialSpecification(), testClass), DEFAULT_TIMEOUT_MS));
        Stream<CTestConfiguration> modelCheckingConfigurations = Arrays.stream(testClass.getAnnotationsByType(ModelCheckingCTest.class))
            .map(ann -> new ModelCheckingCTestConfiguration(testClass, ann.iterations(),
                    ann.threads(), ann.actorsPerThread(), ann.actorsBefore(), ann.actorsAfter(),
                    ann.generator(), ann.verifier(), ann.checkObstructionFreedom(), ann.hangingDetectionThreshold(),
                    ann.invocationsPerIteration(), DEFAULT_GUARANTEES, ann.requireStateEquivalenceImplCheck(),
                    ann.minimizeFailedScenario(),  chooseSequentialSpecification(ann.sequentialSpecification(), testClass), DEFAULT_TIMEOUT_MS));

        return Stream.of(stressConfigurations, modelCheckingConfigurations).flatMap(identity()).collect(Collectors.toList());
    }
}

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/CTestConfiguration.kt
/*
 * Lincheck
 *
 * Copyright (C) 2019 - 2023 JetBrains s.r.o.
 *
 * This Source Code Form is subject to the terms of the
 * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
 * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package org.jetbrains.kotlinx.lincheck

import org.jetbrains.kotlinx.lincheck.CTestConfiguration.Companion.DEFAULT_TIMEOUT_MS
import org.jetbrains.kotlinx.lincheck.execution.*
import org.jetbrains.kotlinx.lincheck.strategy.*
import org.jetbrains.kotlinx.lincheck.strategy.managed.*
import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.*
import org.jetbrains.kotlinx.lincheck.strategy.stress.*
import org.jetbrains.kotlinx.lincheck.transformation.InstrumentationMode
import org.jetbrains.kotlinx.lincheck.verifier.*
import org.jetbrains.kotlinx.lincheck.verifier.linearizability.*
import java.lang.reflect.*

/**
 * Abstract configuration for different lincheck modes.
 */
abstract class CTestConfiguration(
    val testClass: Class<*>,
    val iterations: Int,
    val threads: Int,
    val actorsPerThread: Int,
    val actorsBefore: Int,
    val actorsAfter: Int,
    val generatorClass: Class<out ExecutionGenerator>,
    val verifierClass: Class<out Verifier>,
    val minimizeFailedScenario: Boolean,
    val sequentialSpecification: Class<*>,
    val timeoutMs: Long,
    val customScenarios: List<ExecutionScenario>
) {

    /**
     * Specifies the transformation required for this strategy.
     */
    internal abstract val instrumentationMode: InstrumentationMode

    abstract fun createStrategy(
        testClass: Class<*>, scenario: ExecutionScenario, validationFunction: Actor?,
        stateRepresentationMethod: Method?
    ): Strategy

    companion object {
        const val DEFAULT_ITERATIONS = 100
        const val DEFAULT_THREADS = 2
        const val DEFAULT_ACTORS_PER_THREAD = 5
        const val DEFAULT_ACTORS_BEFORE = 5
        const val DEFAULT_ACTORS_AFTER = 5
        val DEFAULT_EXECUTION_GENERATOR: Class<out ExecutionGenerator?> = RandomExecutionGenerator::class.java
        val DEFAULT_VERIFIER: Class<out Verifier> = LinearizabilityVerifier::class.java
        const val DEFAULT_MINIMIZE_ERROR = true
        const val DEFAULT_TIMEOUT_MS: Long = 3_000_000 // 30 sec
    }
}

internal fun CTestConfiguration.createVerifier() = verifierClass.getConstructor(Class::class.java).newInstance(sequentialSpecification)

internal fun createFromTestClassAnnotations(testClass: Class<*>): List<CTestConfiguration> {
    val stressConfigurations: List<CTestConfiguration> = testClass.getAnnotationsByType(StressCTest::class.java)
        .map { ann: StressCTest ->
            StressCTestConfiguration(
                testClass = testClass,
                iterations = ann.iterations,
                threads = ann.threads,
                actorsPerThread = ann.actorsPerThread,
                actorsBefore = ann.actorsBefore,
                actorsAfter = ann.actorsAfter,
                generatorClass = ann.generator.java,
                verifierClass = ann.verifier.java,
                invocationsPerIteration = ann.invocationsPerIteration,
                minimizeFailedScenario = ann.minimizeFailedScenario,
                sequentialSpecification = chooseSequentialSpecification(ann.sequentialSpecification.java, testClass),
                timeoutMs = DEFAULT_TIMEOUT_MS,
                customScenarios = emptyList()
            )
        }
    val modelCheckingConfigurations: List<CTestConfiguration> =
        testClass.getAnnotationsByType(ModelCheckingCTest::class.java)
            .map { ann: ModelCheckingCTest ->
                ModelCheckingCTestConfiguration(
                    testClass = testClass,
                    iterations = ann.iterations,
                    threads = ann.threads,
                    actorsPerThread = ann.actorsPerThread,
                    actorsBefore = ann.actorsBefore,
                    actorsAfter = ann.actorsAfter,
                    generatorClass = ann.generator.java,
                    verifierClass = ann.verifier.java,
                    checkObstructionFreedom = ann.checkObstructionFreedom,
                    hangingDetectionThreshold = ann.hangingDetectionThreshold,
                    invocationsPerIteration = ann.invocationsPerIteration,
                    guarantees = ManagedCTestConfiguration.DEFAULT_GUARANTEES,
                    minimizeFailedScenario = ann.minimizeFailedScenario,
                    sequentialSpecification = chooseSequentialSpecification(
                        ann.sequentialSpecification.java,
                        testClass
                    ),
                    timeoutMs = DEFAULT_TIMEOUT_MS,
                    customScenarios = emptyList(),
                    stdLibAnalysisEnabled = false,
                )
            }
    return stressConfigurations + modelCheckingConfigurations
}